File: medium/56. Merge Intervals/56. Merge Intervals.py
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda val: val[0])  # [[1, 3], [2, 6], [8, 10], [15, 18]]

        res: List[List[int]] = [intervals[0]]  # [[1, 3]]

        for i in range(1, len(intervals)):  # go through each element
            if intervals[i][0] <= res[-1][1]:  # if 2 <= 3
                res[-1] = [
                    res[-1][0],
                    max(res[-1][1], intervals[i][1]),
                ]  # [1, max(3, 6)]
            else:
                res.append(intervals[i])

        return res


solution = Solution()

print(solution.merge([[1, 3], [2, 6], [8, 10], [15, 18]]))
# print(solution.merge([[1, 4], [4, 5]]))

File: medium/56. Merge Intervals/56. Merge Intervals.ts
function merge(intervals: number[][]): number[][] {
    const sorted_intervals = intervals.sort((a, b) => a[0] - b[0])
    const res: number[][] = [sorted_intervals[0]]
    let resIndex = 0

    for (let i = 1; i < sorted_intervals.length; i++) {
        const lastRes = res[resIndex]
        if (sorted_intervals[i][0] <= lastRes[1]) {
            res[resIndex] = [res[resIndex][0], Math.max(res[resIndex][1], sorted_intervals[i][1])]
        } else {
            res.push(sorted_intervals[i])
            resIndex += 1
        }
    }

    return res
};